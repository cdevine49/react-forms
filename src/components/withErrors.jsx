import React from 'react';

const withErrors = Input => class extends React.Component {
  constructor() {
    super();
    this.state = {
      errorMessage: '',
      displayErrors: false
    }
    this._hideErrors = this._hideErrors.bind(this);
    this._displayErrors = this._displayErrors.bind(this);
    this._handleErrors = this._handleErrors.bind(this);
  }

  componentDidMount() {
    // this.props.validate({[this.props.id]: !this.props.required});
    this._handleErrors(this.props.value);
  }

  componentWillReceiveProps(nextProps) {
    !this.props.displayErrors && nextProps.displayErrors && this.setState({ displayErrors: true })
  }

  _handleErrors(value) {
    let errorMessage = false;
    if (this.props.required && !value) {
      errorMessage = "You can't leave this empty";
    } else if (value && this.props.errors) {
      for (var i = 0; i < this.props.errors.length; i++ ) {
        var error = this.props.errors[i];
        if (error._handle(value)) {
          errorMessage = error.message;
        }
      }
    }


    this.setState({ errorMessage });
    // this.props.validate({[this.props.id]: !errorMessage})
  }

  _displayErrors() {
    this.setState({ displayErrors: true });
  }

  _hideErrors() {
    this.setState({ displayErrors: false });
  }

  render() {
    return <Input handleErrors={this._handleErrors} onBlur={this._displayErrors} onFocus={this._hideErrors} {...this.props} {...this.state} />;
  }
}

export default withErrors;