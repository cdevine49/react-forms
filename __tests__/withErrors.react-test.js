import React from 'react';
import withErrors from '../src/components/withErrors'
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';

describe('WithErrors', () => {
  let Root, wrapper, WrappedComponent, message = 'test message';
  beforeEach(() => {
    Root = () => {
      return (
        <div id="root" className="return element"></div>
      );
    }
    WrappedComponent = withErrors(Root);
    wrapper = shallow(<WrappedComponent errors={[{_handle: () => true, message}]} />);
  });

  it('renders the passed component as its root element', () => {
    const tree = renderer.create(
      <WrappedComponent
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('passes errorMessage state as prop to its root element', () => {
    const message = 'test message';
    expect(wrapper.first().props().errorMessage).toBe('');
    wrapper.setState({ errorMessage: message });
    expect(wrapper.first().props().errorMessage).toBe(message);
  });

  it('passes onFocus and onBlur functions as props to its root element', () => {
    expect(typeof wrapper.first().props().onFocus).toBe('function');
    expect(typeof wrapper.first().props().onBlur).toBe('function');
  });

  it('sets errorMessage to message of error that needs handing onBlur', () => {
    wrapper.first().props().onBlur();
    expect(wrapper.first().props().errorMessage).toBe(message);
  });

  it('sets errorMessage to empty onFocus', () => {
    wrapper.setState({ errorMessage: message });
    wrapper.first().props().onFocus();
    expect(wrapper.first().props().errorMessage).toBe('');
  });
});