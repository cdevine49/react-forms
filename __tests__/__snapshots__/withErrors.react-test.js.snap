// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithErrors renders the passed component as its root element 1`] = `
<div
  className="return element"
  id="root"
/>
`;
